tfsdk/block.go

```diff
@@ -0,0 +1,406 @@
+}
+
+// tfprotov6 returns the *tfprotov6.SchemaNestedBlock equivalent of a Block.
+// Errors will be tftypes.AttributePathErrors based on `path`. `name` is the
+// name of the attribute.
+func (b Block) tfprotov6(ctx context.Context, name string, path *tftypes.AttributePath) (*tfprotov6.SchemaNestedBlock, error) {
+	schemaNestedBlock := &tfprotov6.SchemaNestedBlock{
+		Block: &tfprotov6.SchemaBlock{
+			Deprecated: b.DeprecationMessage != "",
+		},
+		MinItems: b.MinItems,
+		MaxItems: b.MaxItems,
+		TypeName: name,
+	}
+
+	if b.Description != "" {
+		schemaNestedBlock.Block.Description = b.Description
+		schemaNestedBlock.Block.DescriptionKind = tfprotov6.StringKindPlain
+	}
+
+	if b.MarkdownDescription != "" {
+		schemaNestedBlock.Block.Description = b.MarkdownDescription
+		schemaNestedBlock.Block.DescriptionKind = tfprotov6.StringKindMarkdown
+	}
+
+	nm := b.NestingMode
+	switch nm {
+	case BlockNestingModeList:
+		schemaNestedBlock.Nesting = tfprotov6.SchemaNestedBlockNestingModeList
+	case BlockNestingModeSet:
+		schemaNestedBlock.Nesting = tfprotov6.SchemaNestedBlockNestingModeSet
+	default:
+		return nil, path.NewErrorf("unrecognized nesting mode %v", nm)
+	}
+
+	for attrName, attr := range b.Attributes {
+		attrPath := path.WithAttributeName(attrName)
+		attrProto6, err := attr.tfprotov6SchemaAttribute(ctx, attrName, attrPath)
+
+		if err != nil {
+			return nil, err
+		}
+
+		schemaNestedBlock.Block.Attributes = append(schemaNestedBlock.Block.Attributes, attrProto6)
+	}
+
+	for blockName, block := range b.Blocks {
+		blockPath := path.WithAttributeName(blockName)
+		blockProto6, err := block.tfprotov6(ctx, blockName, blockPath)
+
+		if err != nil {
+			return nil, err
+		}
+
+		schemaNestedBlock.Block.BlockTypes = append(schemaNestedBlock.Block.BlockTypes, blockProto6)
+	}
+
+	sort.Slice(schemaNestedBlock.Block.Attributes, func(i, j int) bool {
+		if schemaNestedBlock.Block.Attributes[i] == nil {
+			return true
+		}
+
+		if schemaNestedBlock.Block.Attributes[j] == nil {
+			return false
+		}
+
+		return schemaNestedBlock.Block.Attributes[i].Name < schemaNestedBlock.Block.Attributes[j].Name
+	})
+
+	sort.Slice(schemaNestedBlock.Block.BlockTypes, func(i, j int) bool {
+		if schemaNestedBlock.Block.BlockTypes[i] == nil {
+			return true
+		}
+
+		if schemaNestedBlock.Block.BlockTypes[j] == nil {
+			return false
+		}
+
+		return schemaNestedBlock.Block.BlockTypes[i].TypeName < schemaNestedBlock.Block.BlockTypes[j].TypeName
+	})
+
+	return schemaNestedBlock, nil
+}
+
+// validate performs all Block validation.
+func (b Block) validate(ctx context.Context, req ValidateAttributeRequest, resp *ValidateAttributeResponse) {
+	attributeConfig, diags := req.Config.getAttributeValue(ctx, req.AttributePath)
+	resp.Diagnostics.Append(diags...)
+
+	if diags.HasError() {
+		return
+	}
+
+	req.AttributeConfig = attributeConfig
+
+	for _, validator := range b.Validators {
+		validator.Validate(ctx, req, resp)
```

Comment:
I forget offhand, but do we need to return early here if `resp.Diagnostics.HasError`?

Reply:
No worries! The current implementation of `(Attribute).validate()` loops through `Validators` without skipping early:

https://github.com/hashicorp/terraform-plugin-framework/blob/16188d2b2de27dc006107f5d637aa9d7e44b7a37/tfsdk/attribute.go#L274-L276

The rationale that I can recall is that since the configuration value is not changing during this process, that all validators should return their errors as they likely will represent different expectations, e.g.

```go
Validators: []AttributeValidator{
  stringLengthAtMost(8),
  stringBeginsWith("tf"),
},
```

So if `some-long-string-without-tf-prefix!` was passed in, both validation error diagnostics would be returned. 👍

Reply:
That seems reasonable to me!